feat(compiler-cli): support multiple configuration files in `extends`

alan-agius4 opened the pull request:
TypeScript 5 support `extends` to be an array, this commit adds support to allow extending `angularCompilerOptions` from multiple config files.

See: https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#supporting-multiple-configuration-files-in-extends

alan-agius4 commented:
Blocked on https://github.com/angular/angular/pull/49107

alan-agius4 committed:
feat(compiler): support multiple configuration files in `extends`

TypeScript 5 support `extends` to be an array, this commit adds support to allow extending `angularCompilerOptions` from multiple config files.

See: https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#supporting-multiple-configuration-files-in-extends

JoostK commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -194,4 +194,52 @@ describe('perform_compile', () => {
        }));
      });
+
+  it('should merge tsconfig "angularCompilerOptions" when extends is an array', () => {
```
I wonder what the behavior here should be for nested options, `extendedDiagnostics` being the primary example.

alan-agius4 replied:
That doesn’t change with this change. Complex objects such as `paths` and `extendedDiagnostics` ste not merged.

JoostK replied:
Do we have a test to that effect? I realise it's more about `extends` in general than having multiple `extends` per se, but wouldn't hurt to explicitly capture this behavior in a test.

alan-agius4 replied:
We do not, but I can add it if you wish.

JoostK replied:
Capturing that behavior would be valuable, IMO

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -79,25 +80,34 @@ export function readConfiguration(
+
+          // Call readAngularCompilerOptions recursively to merge NG Compiler options
+          // Reserve the array so the override overrides happen from right to left.
+          return extendsPaths.reverse().reduce((prevOptions, extendsPath) => {
```
This ends up reversing `extendsPath` in place, which could become a problem e.g. if `readConfigFile` were cached.

alan-agius4 replied:
Good caych

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -79,25 +80,34 @@ export function readConfiguration(
+              typeof config.extends === 'string' ? [config.extends] : config.extends;
+
+          // Call readAngularCompilerOptions recursively to merge NG Compiler options
+          // Reserve the array so the override overrides happen from right to left.
```
Reserve -> Reverse

alan-agius4 committed:
fixup! feat(compiler): support multiple configuration files in `extends`

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -79,25 +80,34 @@ export function readConfiguration(
+              typeof config.extends === 'string' ? [config.extends] : config.extends;
+
+          // Call readAngularCompilerOptions recursively to merge NG Compiler options
+          // Reverse the array so the override overrides happen from right to left.
```
I'm now noticing some duplication here:

```suggestion
          // Reverse the array so the overrides happen from right to left.
```

Also this would be clearer if the prior sentence ends with a dot.

alan-agius4 committed:
fixup! feat(compiler): support multiple configuration files in `extends`

alan-agius4 committed:
test(compiler): add test case for merging objects in tsconfig

This commit adds a test case for validate that we do not deep merge objects like like 'paths' and `extendedDiagnostics`.

jessicajaniuk commented:
This PR was merged into the repository by commit 1418d1937fd779afb6e23b46143f3fb7e0ddcfd1.
